### PR TITLE
feat(storagenode): change the synchronization method to accept only the last commit context

### PR DESCRIPTION
**Pull request by ijsong**
### What this PR does

This patch changes the synchronization process to accept only the last commit context applied by #162.
The synchronization copies log entries and the latest commit context from the source replica to the
destination replica. The destination's reportCommitBase should be valid, and its log entries should
be the same as the source's.

SyncInit can now delete log entries that the source replica had already removed. It deletes all log
entries simultaneously to avoid a hole in the middle of log entries. It also makes the replica's
reportCommitBase invalid because commit context and log entry are inconsistent during the
synchronization.

SyncReplicate no longer uses the SyncReplicateBuffer since there is no reason to store a commit
context and corresponding log entries in the buffer and write them into a disk at once. It now saves
log entries or the commit context every time. We can optimize the SyncReplicate to transfer a batch
of log entries, which we can discuss later.

### Which issue(s) this PR resolves

Updates #125

### CODE REVIEW

File: internal/storagenode/storagenode_test.go [hunk 1]
```diff
@@ -581,3 +590,476 @@ func TestStorageNode_Report(t *testing.T) {
...
+		}
+	}
+	trim := func(t *testing.T, sn *StorageNode, trimGLSN types.GLSN) {
+		ret := TestTrim(t, cid, sn.snid, tpid, trimGLSN, sn.advertise)
+		require.Len(t, ret, 1)
+		require.Contains(t, ret, lsid)
+		require.NoError(t, ret[lsid])
+	}
+	_ = put
+	_ = trim
```
Comment (hungryjang) [inline internal/storagenode/storagenode_test.go:664]: What does this do?

File: internal/storagenode/storagenode_test.go [hunk 2]
```diff
@@ -581,3 +590,476 @@ func TestStorageNode_Report(t *testing.T) {
...
+				snmd, err := dst.getMetadata(context.Background())
+				require.NoError(t, err)
+				require.Equal(t, varlogpb.LogSequenceNumber{
+					LLSN: 4, GLSN: 4,
+				}, snmd.LogStreamReplicas[0].LocalLowWatermark)
+				require.Equal(t, varlogpb.LogSequenceNumber{
+					LLSN: 4, GLSN: 4,
+				}, snmd.LogStreamReplicas[0].LocalHighWatermark)
+			},
+		},
```
Comment (hungryjang) [inline internal/storagenode/storagenode_test.go:1019]: How about adding this case?

ver: +-1-+ +-2-+ +-3-+
src:  ............3 4......5 6
dst:  ...1 2....3 4

File: internal/storagenode/logstream/sync.go [hunk 3]
```diff
@@ -386,34 +268,22 @@ func (lse *Executor) SyncInit(_ context.Context, srcReplica varlogpb.LogStreamRe
 		return
 	}
 
-	trimmed := uncommittedLLSNBegin < srcRange.FirstLLSN
+	trimmed := lastCommittedLLSN < srcRange.FirstLLSN
```
[the remark this answers is not in the document]
Comment (ijsong) [reply internal/storagenode/logstream/sync.go:271]: Yes, it is. The variable `trimmed` declared here indicates whether all log entries in a destination replica are trimmed. 
However, I agree that we can trim partially, as you said. I will modify this soon at this PR.

File: internal/storagenode/storagenode_test.go (hunk 1 above)
Comment (ijsong) [reply internal/storagenode/storagenode_test.go:664]: Uh, I forgot to delete those lines.

File: internal/storagenode/storagenode_test.go (hunk 2 above)
Comment (ijsong) [reply internal/storagenode/storagenode_test.go:1019]: Good, this is the case you pointed out.
According to the current implementation, the destination has log entries from 1 to 6 after synchronization. If we consider partial trimming, the destination will have log entries from 3 to 6.

File: internal/storagenode/storagenode_test.go (hunk 1 above)
Comment (ijsong) [reply internal/storagenode/storagenode_test.go]: Deleted.

File: internal/storagenode/storagenode_test.go (hunk 2 above)
Comment (ijsong) [reply internal/storagenode/storagenode_test.go:1015]: Added this test case.

File: internal/storagenode/logstream/sync.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ijsong) [reply internal/storagenode/logstream/sync.go]: Modified. Related test case:

https://github.com/ijsong/varlog/blob/f89e6ca4de260f5b9b5e89af642640c0a3f34038/internal/storagenode/storagenode_test.go#L1016-L1059

```go
			// ver: +-1-+ +-2-+ +-3-+
			// src:        3 4   5 6
			// dst:  1 2   3 4
```

**Review by hungryjang — APPROVED** [review]

**Pull request merged**